### PR TITLE
[dbx-num] add state machine to 526

### PR DESCRIPTION
**Pull request by SamStuckey**
## Summary
*NOTE: this is PR 3 of 3 to add this logic*
- [PR 1](https://github.com/department-of-veterans-affairs/vets-api/pull/15736)
- [PR 2](https://github.com/department-of-veterans-affairs/vets-api/pull/15801)

### The Problem
We are about to do an audit in cooperation with VBMS to determine the success of our past year of remediation efforts.  Though VBMS is the source of truth, we need a way to easily mark submissions as conceptually 'done' or 'off our plate'.  

To date, we have been using the presence of `submitted_claim_id` and `backup_submitted_claim_id` to represent success via the primary or backup submission path, respectively.  

**This did not account for our 'failsafe' remediations**

Additionally, with a simple state machine in place we have a way to record the results of Lighthouse's status updates. Even after we send a submission to lighthouse, it isn't "off our plate" until LH has marked it as successful (status of `vbms`).

### The solution

- Add a basic state machine to the form 526 Submission flow with pass fail states for VBMS and remediation
- Add other transitory states for completeness

### Other stuff
- add state changes everywhere it's applicable in our current code.  This covers only a few of the available states, as it will take a while to backfill states (as part of our audit) and to implement our status polling (will inform state transitions).

## Omissions
This work *does not* add any transition validation to the state machine.  That would be out of scope for MVP, and dangerous to roll out as it could cause errors in the submission flow.  MVP is to simply add the datapoint, unblocking our audit of VBMS.

## Related issue(s)

- [Ticket](https://app.zenhub.com/workspaces/disability-benefits-experience-team-carbs-6470c8bfffee9809b2634a52/issues/gh/department-of-veterans-affairs/va.gov-team/75526)
- [Feature / Implementation Documentation](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/implementation/form_526_state_machine.md)

## Testing done

- [x] *New code is covered by unit tests*
- Instantiation of affected objects and testing for continuity

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
- Form 526, but not in a veteran facing way.  this will be used primarily for remediation of failed submissions, at least for MVP

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: db/migrate/20240123221602_add_aasm_state_to_form_526_submissions.rb
Comment (tblackwe) [inline db/migrate/20240123221602_add_aasm_state_to_form_526_submissions.rb:1]: This needs to be separated out to its own PR. I believe this is because of the automated job that runs on migrations
https://depo-platform-documentation.scrollhelp.site/developer-docs/vets-api-database-migrations#vets-apidatabasemigrations-Deployment

**Review by tblackwe — CHANGES_REQUESTED** [review]

**Comment by SamStuckey** [comment]
Breaking this out into 3 new PRs per the docs.  First one is here
https://github.com/department-of-veterans-affairs/vets-api/pull/15736
adding the new state column to the table

File: app/models/form526_submission.rb [hunk 1]
```diff
@@ -6,10 +6,82 @@
...
+
+    # - a successfully delivered submission has failed 3rd party validations on primary path
+    # - requires backup submission or remediation
+    event :reject_from_primary do
+      transitions to: :rejected_by_primary
+    end
+
+    # - a submission has been delivered to our backup path
+    # - accepts any non-success type state
+    # - MUST accept all the same incomming statuses as fail_backup_delivery
```
Comment (jerekshoe) [inline app/models/form526_submission.rb:43]: Minor, but "incomming" should be "incoming". There is also another instance of this misspelling on L51 as well

File: app/models/form526_submission.rb [hunk 2]
```diff
@@ -6,10 +6,82 @@
...
+
+    state :unprocessed, initial: true
+    state :delivered_to_primary, :failed_primary_delivery, :rejected_by_primary,
+          :delivered_to_backup, :failed_backup_delivery, :rejected_by_backup,
+          :in_remediation, :finalized_as_successful, :unprocessable
+
+    # - a submission has been delivered to our happy path
+    # - requires polling to finalize
+    event :deliver_to_primary do
+      transitions to: :delivered_to_primary
```
Comment (freeheeling) [inline app/models/form526_submission.rb:26]: Should the state from which an event transitions be declared, or is that not necessary as it's always from unprocessed?

File: app/models/form526_submission.rb [hunk 3]
```diff
@@ -6,10 +6,82 @@
...
+
+    # - a successfully delivered submission has failed 3rd party validations on primary path
+    # - requires backup submission or remediation
+    event :reject_from_primary do
+      transitions to: :rejected_by_primary
+    end
+
+    # - a submission has been delivered to our backup path
+    # - accepts any non-success type state
+    # - MUST accept all the same incoming statuses as fail_backup_delivery
```
Comment (freeheeling) [inline app/models/form526_submission.rb:43]: This is the same comment as on L51. Will you explain why the `deliver_to_backup` event MUST accept all the same incoming statuses as `fail_backup_delivery`?

File: app/models/form526_submission.rb [hunk 4]
```diff
@@ -6,10 +6,82 @@
...
+
+    # - Submission has entered a manual remediation flow, e.g. failsafe, paper submission
+    # - requires confirmation of success, e.g. polling or manual confirmation via audit
+    event :begin_remediation do
+      transitions to: :in_remediation
+    end
+
+    # - The only state that means we no longer own completion of this submission
+    # - There is nothing more to do.  E.G.
+    #   - VBMS has accepted and returned the applicable status to us via LH
```
Comment (freeheeling) [inline app/models/form526_submission.rb:71]: Suggest writing out Lighthouse here, to avoid any future confusion regarding meaning.

File: lib/sidekiq/form526_backup_submission_process/submit.rb [hunk 5]
```diff
@@ -29,6 +29,7 @@ class Submit
         error_message = msg['error_message']
         timestamp = Time.now.utc
         form526_submission_id = msg['args'].first
+        Form526Submission.find_by(id: form526_submission_id).fail_backup_delivery!
```
Comment (freeheeling) [inline lib/sidekiq/form526_backup_submission_process/submit.rb:32]: `Form526Submission.find(form526_submission_id)` might be more appropriate if you expect the record to always exist. Otherwise, I suggest adding a safe navigation operator. Same on L83.

File: spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb [hunk 6]
```diff
@@ -30,5 +30,22 @@
         expect { subject.new.perform(submission.id) }.to raise_error NotImplementedError
       end
     end
+
+    context 'when all retrys are exhausted' do
```
Comment (freeheeling) [inline spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:34]: 'retrys' should be 'retries'

File: app/models/form526_submission.rb [hunk 7]
```diff
@@ -6,10 +6,82 @@
...
+      transitions to: :delivered_to_backup
+    end
+
+    # - a submission has failed to be delivered to our backup path
+    # - accepts any non-success type state
+    # - MUST accept all the same incoming statuses as fail_backup_delivery
+    # - requires remediation
+    event :fail_backup_delivery do
+      transitions to: :failed_backup_delivery
+    end
```
Comment (freeheeling) [inline app/models/form526_submission.rb:55]: Is there a test for this transition?

Comment (SamStuckey) [reply app/models/form526_submission.rb:53]: added, plus a few others that i missed

File: app/models/form526_submission.rb (hunk 2 above)
Comment (SamStuckey) [reply app/models/form526_submission.rb:26]: typically yes, however (i think this is the doc) this is not a true state machine.  we are allowing any state transition for MVP.  This is intentional to 
1. avoide unintended concequences
2. allow us freedom in remediation, which can be unpredictable.
IF / when this becomes a real state machine we can add that

File: app/models/form526_submission.rb (hunk 3 above)
Comment (SamStuckey) [reply app/models/form526_submission.rb:42]: Those comments are outdated; removing them.  check the docs if you want to better grok to actual lifecycle flows that will (probalby) happen.  Again, we are allowing uncertainty by not enforcing `from`  state because the primary value here is dog-earing remediated submissions.

**Review by freeheeling — DISMISSED** [review]
:shipit:

**Review by tblackwe — APPROVED** [review]

**Comment by SamStuckey** [comment]
Marking Ready for review with an outstanding linting error on my test.  The linting concern is about the stubbing of a method provided by an included module.  The reason I can't (at least I haven't figured out how) to fix this, is that `within_sidekiq_retries_exhausted_block` doesn't recognize *either* the JobTracking module that defines if *or* the parent class into which the JobTracking module is composed as available.

File: spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb [hunk 8]
```diff
@@ -30,5 +30,22 @@
         expect { subject.new.perform(submission.id) }.to raise_error NotImplementedError
       end
     end
+
+    context 'when all retries are exhausted' do
+      let!(:form526_submission) { create(:form526_submission) }
+      let!(:form526_job_status) { create(:form526_job_status, :non_retryable_error, form526_submission:, job_id: 1) }
+
+      it 'transitions the submission to a failure state' do
+        allow(subject).to receive(:job_exhausted)
```
Comment (LindseySaari) [inline spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:39]: Would something like `subject(:disability_comp_submit_for_526_job) { EVSS::DisabilityCompensationForm::SubmitForm526.new.job_exhausted }` work here?

Comment (SamStuckey) [reply spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:39]: I'm not 100% clear on what you mean?  Are you proposing to call the method manually without using the `within_sidekiq_retries_exhausted_block` helper?  If so (ie. `EVSS::DisabilityCompensationForm::SubmitForm526.sidekiq_retries_exhausted` this will not actually execute the block.  We might be able to finagle something like that, but imho if it's that important to test without stubbing this, I think the "right way" would be to figure out the RSpec configuration required to include module logic in the test scope.

Can you elaborate a bit?

Comment (LindseySaari) [reply spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:41]: I'm just spitballing ideas to get the Linting to pass

Comment (LindseySaari) [reply spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:41]: Was taking a look at the documentation for rubocop here: https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/SubjectStub

Comment (SamStuckey) [reply spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:41]: Gottcha, thanks for following up.  So I get why it's failing, the code is in clear violation of the policy.  I'm not saying it's a bad failure or that it doesn't make sense; I'm proposing ignoring what is generally a good rule in this case because of the aforementioned reasons. : )

But, if that feels wrong, I'll just dig into the RSpec config.  It's not that I don't know how to get the test to pass, it's that I don't know if it's worth the amount of work given the context. How does that sit?  Honestly I'm happy to fix it, i just thought this seemed like a fair case for  and exception, just need to know what the platform team prefers.

Comment (SamStuckey) [reply spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb:41]: I'll just fix it, no sweat.  Thanks for the feedback.

Comment (SamStuckey) [reply spec/sidekiq/evss/disability_compensation_form/submit_form526_spec.rb]: @LindseySaari all fixed up.  Rebased for good measure, thanks again : )

**Review by tblackwe — APPROVED** [review]

**Review by LindseySaari — APPROVED** [review]
